### ISSUE TITLE
Update handler to follow best practices

### ISSUE THREAD

**Issue by F21** [neelance/graphql-go#36]
It would be really cool if `relay.Handler` can be updated to support the best practices outlined here: http://graphql.org/learn/serving-over-http/

For example, it would be nice to support both `GET` and `POST` and deal with different `Content-Type`s.

I think the handler from `graphql-go` can be used as a reference as it is pretty complete: https://github.com/graphql-go/handler/blob/master/handler.go

**Comment by neelance** [comment]
The `relay` package is meant as a backend for Relay, which only needs the functionality currently implemented. However, I see that it is likely used as a starting point for a GraphQL HTTP endpoint. I'm willing to consider adding those best practices, since they are official enough. I'd just like to first know some real world example on how those additional ways are going to be used. What made you ask for this? :)

**Comment by nicksrandall** [comment]
One technique that I am using that is baked into apollo client (popular js graphql client) is [query batching](https://dev-blog.apollodata.com/query-batching-in-apollo-63acfd859862#.lbhojacqr).

TLDR; My graphql endpoint accepts a query or an array of queries. When an array of queries is received, we return an array of responses.

My handler looks something like this: 

```go
type GraphqlHandler struct {
	Schema *graphql.Schema
	pretty bool
}

type RequestOptions struct {
	Query         string                 `json:"query"`
	Variables     map[string]interface{} `json:"variables"`
	OperationName string                 `json:"operationName"`
}

func (h *GraphqlHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	var output interface{}
	var b []byte
	var responseJSON []byte
	var err error

	b, err = ioutil.ReadAll(r.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	if bytes.HasPrefix(b, []byte("[")) {
		var responses []*graphql.Response
		var queries []*RequestOptions
		json.Unmarshal(b, &queries)
		for i := range queries {
			responses = append(responses, h.Schema.Exec(r.Context(), queries[i].Query, queries[i].OperationName, queries[i].Variables))
		}
		output = responses
	} else {
		var query RequestOptions
		json.Unmarshal(b, &query)
		output = h.Schema.Exec(r.Context(), query.Query, query.OperationName, query.Variables)
	}

	if h.pretty {
		responseJSON, err = json.MarshalIndent(output, "", "  ")
	} else {
		responseJSON, err = json.Marshal(output)
	}

	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Write(responseJSON)
}
```

**Comment by neelance** [comment]
Query batching will not become part of the `relay` package until Relay officially supports it. But why don't you release an `apollo` package? ;-)

**Comment by nicksrandall** [comment]
Are you suggesting that I add a new package in this repo or a new repo all together? I'm happy to contribute either way.

**Comment by neelance** [comment]
Let's have a new repo. We can add a link to the readme. The `relay` package may also go into a new repo at some point.

**Comment by F21** [comment]
@neelance Initially I was looking at the `graphql-go/graphql` package, but it seems to be somewhat unmaintained. In my case, I would like to build a few graphql services for a few simple internal APIs and don't really need the extra functionality provided by Relay and having an official handler would make it really easy to start up a graphql server 😄

**Comment by nicksrandall** [comment]
@neelance here is my apollo handler: https://github.com/nicksrandall/apollo-go/blob/master/handler.go if you want to link to it?

**Comment by neelance** [comment]
@F21 Is @nicksrandall's project what you are looking for?

**Comment by F21** [comment]
@neelance Yes, pretty much! @nicksrandall would you guys be interested in moving the handler into this project?

**Comment by nicksrandall** [comment]
@F21 we discussed that earlier in this thread. @neelance thought it would be better in it's own package.

**Comment by F21** [comment]
Ah I see. The reason I asked was that the relay package in this repo has a handler, so if we can compose it with your handler, the relay handler will get access to those best practices for free.

**Comment by 0xCMP** [comment]
The changes @nicksrandall has made break the example graphiql server because it doesn't support the `"text/plain"` encoding.

I was fine with the relay handler, but it doesn't handle option requests gracefully which is causing my `apollo-client` code to error. I could simply wrap it to fix the problem, but I think it'd be better to improve either one of the handlers to handle options/head/etc. methods properly.

Thoughts?

**Comment by nicksrandall** [comment]
If you can show me where "text/plain" is officially supported, then I will add it.

**Comment by 0xCMP** [comment]
Well, while my main issues aren't "text/plain", and I don't think I've seen it officially supported anywhere, this graphiql client [0] (using version 0.7.8, current being 0.8) does send "text/plain" for it's initial request.

My issues had more to do with handling CORS requests [1]. I can make this another issue if @neelance wants since I think it's looking to be a bit more off-topic than I thought it'd be 

(0): https://github.com/neelance/graphql-go/blob/8f7d2b1efd9e96e423c0dc0ecf7dbc905e1b5ff0/example/starwars/server/server.go
(1): https://fetch.spec.whatwg.org/#cors-preflight-fetch

**Comment by 0xCMP** [comment]
Having understood the problem better, I think this isn't related to your handlers. Probably better to handle myself than make either handlers worry about this.

**Comment by neelance** [comment]
I'm still not convinced that the `relay` handler needs more than Relay actually uses. Closing for now, but I'm still open to discussion.

**Comment by prabhatga** [comment]
To begin with: thanks for a fantastic package.

Is it possible to provide for a functionality in relay.Handler to set additional custom headers? One use case is CORS. Let's say one wants to set some custom headers while sending an HTTP response. Wrapping relay.Handler and setting the headers in the wrapper will not work -- since the request body is being read in ServeHTTP, and setting the headers in the wrapper throws the error ""http: invalid Read on closed Body".

The solution to write a custom version of ServeHTTP with support for adding custom headers, but I was wondering if you think it's a feature that makes sense in relay.Handler.

Would be interested in knowing your thoughts on this. Thanks.

**Comment by bsr203** [comment]
This is the way I do it now.

```
s, err := schema.ResolvedSchema()
	if err != nil {
		panic(err)
	}
    
http.Handle("/graphql", NewGraphqlHandler(s))


type Handler struct {
	Schema *graphql.Schema
}

// NewGraphqlHandler creates an new GraphQL API HTTP handler with the specified graphql schema
func NewGraphqlHandler(s *graphql.Schema) *Handler {
	return &Handler{Schema: s}
}


func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	// dump, err := httputil.DumpRequest(r, true)
	// if err != nil {
	// 	http.Error(w, fmt.Sprint(err), http.StatusInternalServerError)
	// 	return
	// }
	// fmt.Printf("%s", dump)

	if origin := r.Header.Get("Origin"); origin != "" {
		w.Header().Set("Access-Control-Allow-Origin", origin)
	}
	w.Header().Set("Access-Control-Allow-Methods", "POST, GET, OPTIONS, HEAD")

	if r.Method == http.MethodOptions {
		w.Header().Set("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept, Authorization")
		return
	}
	w.Header().Set("Access-Control-Allow-Headers", "Content-Type, Content-Length, Accept-Encoding")

	var params struct {
		Query         string                 `json:"query"`
		OperationName string                 `json:"operationName"`
		Variables     map[string]interface{} `json:"variables"`
	}
	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

    ......
    ......

    response := h.Schema.Exec(ctx, params.Query, params.OperationName, params.Variables)
	responseJSON, err := json.Marshal(response)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Write(responseJSON)
}

```

**Comment by prabhatga** [comment]
Yes, thanks for the suggestion. I am doing something in my own code as well.
